bazel/wasm/wasm.bzl

```diff
@@ -0,0 +1,93 @@
+#  Copyright 2020 Google LLC
+# 
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+# 
+#       http://www.apache.org/licenses/LICENSE-2.0
+# 
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+load("@rules_cc//cc:defs.bzl", "cc_binary")
+
+def _wasm_cc_transition_impl(settings, attr):
+    return {
+        "//command_line_option:cpu": "wasm32",
+        "//command_line_option:crosstool_top": "@proxy_wasm_cpp_sdk//toolchain:emscripten",
+
+        # Overriding copt/cxxopt/linkopt to prevent sanitizers/coverage options leak
+        # into Wasm build configuration
+        "//command_line_option:copt": [],
+        "//command_line_option:cxxopt": [],
+        "//command_line_option:linkopt": [],
+        "//command_line_option:collect_code_coverage": "false",
+        "//command_line_option:fission": "no",
+    }
+
+wasm_cc_transition = transition(
+    implementation = _wasm_cc_transition_impl,
+    inputs = [],
+    outputs = [
+        "//command_line_option:cpu",
+        "//command_line_option:crosstool_top",
+        "//command_line_option:copt",
+        "//command_line_option:cxxopt",
+        "//command_line_option:fission",
+        "//command_line_option:linkopt",
+        "//command_line_option:collect_code_coverage",
+    ],
+)
+
+def wasm_binary_impl(ctx):
+    out = ctx.actions.declare_file(ctx.label.name)
+    ctx.actions.run(
+        executable = "cp",
+        arguments = [ctx.files.binary[0].path, out.path],
+        outputs = [out],
+        inputs = ctx.files.binary,
+    )
+
+    return [DefaultInfo(files = depset([out]), runfiles = ctx.runfiles([out]))]
+
+def _wasm_attrs(transition):
+    return {
+        "binary": attr.label(mandatory = True, cfg = transition),
+        "_whitelist_function_transition": attr.label(default = "@bazel_tools//tools/whitelists/function_transition_whitelist"),
+    }
+
+# Wasm binary rule implementation.
+# This copies the binary specified in binary attribute in Wasm configuration to
+# target configuration, so a binary in non-Wasm configuration can depend on them.
+wasm_cc_binary_rule = rule(
+    implementation = wasm_binary_impl,
+    attrs = _wasm_attrs(wasm_cc_transition),
+)
+
+def wasm_cc_binary(name, tags = [], repository = "", **kwargs):
+    wasm_name = "_wasm_" + name
+    kwargs.setdefault("additional_linker_inputs", ["@proxy_wasm_cpp_sdk//:jslib"])
+    if repository == "@proxy_wasm_cpp_sdk":
+        js = "--js-library proxy_wasm_intrinsics.js"
+    else:
+        js = "--js-library external/proxy_wasm_cpp_sdk/proxy_wasm_intrinsics.js"
```

Comment:
This should be the other way around.

Reply:
For extern repo, they no longer need this hack post 1.8 since js glue is not needed. If we revert the logic here, all extern repo will need to have `repository` specified in wasm build target, which is going to be broken once they update to new SDK SHA since `repository` is not there in master. So I'd rather keep it as is and specify repository only in SDK repo.

Reply:
If repository is confusing here, we can use some other arg name? `external_js_library` to make it more clear.

Reply:
This is how external repositories should use this:
```
wasm_cc_library(
    srcs = [...],
    repository = "@proxy_wasm_cpp_sdk",
    deps = [
        "@proxy_wasm_cpp_sdk//:proxy_wasm_intrinsics",
    ],
)
```
meaning that `.js` glue should be pulled from `@proxy_wasm_cpp_sdk`, i.e. `external/proxy_wasm_cpp_sdk/...`.

The logic in this PR is inverted (at least compared to Envoy rules), and requires this repo (and only this repo) to use `repository = "@proxy_wasm_cpp_sdk"` to mean "use local file".

Changing name would require this repo (and only this repo) to use `external_js_library = "@proxy_wasm_cpp_sdk"` to mean "use local file", which would be even more confusing, IMHO.

Reply:
All I want is to to make external repo usage like
```
wasm_cc_library(
    srcs = [...],
    deps = [
        "@proxy_wasm_cpp_sdk//:proxy_wasm_intrinsics",
    ],
)
```

since `repository` is not going to be needed in the future. How about we make it `js_library_path` and make it by default point to `external/proxy_wasm_cpp_sdk/proxy_wasm_intrinsics.js`. This way only this repo will use this param and point to the `local file` path.

Reply:
OK, that's a valid argument. Maybe just call it `in_primary_repo = true|false` and default to `false`?

Reply:
Done.